app/repositories/onlinetesting/OnlineTestEvaluationRepository.scala

```diff
@@ -0,0 +1,149 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package repositories.onlinetesting
+
+import model.ApplicationStatus._
+import model.Exceptions.PassMarkEvaluationNotFound
+import model.persisted._
+import model.{ ApplicationStatus, ProgressStatuses, SelectedSchemes }
+import reactivemongo.api.DB
+import reactivemongo.bson.{ BSONArray, BSONDocument, BSONObjectID }
+import repositories.{ BSONHelpers, CommonBSONDocuments, RandomSelection }
+import uk.gov.hmrc.mongo.ReactiveRepository
+import uk.gov.hmrc.mongo.json.ReactiveMongoFormats
+import model.Phase._
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+trait OnlineTestEvaluationRepository[T] extends CommonBSONDocuments with BSONHelpers with RandomSelection {
+
+  this: ReactiveRepository[_, _] =>
+
+  val phase: Phase
+
+  val evaluationApplicationStatuses: List[ApplicationStatus]
+
+  val applicationBSONReader = (doc: BSONDocument) => {
+    val applicationId = doc.getAs[String]("applicationId").get
+    val applicationStatus = doc.getAs[ApplicationStatus]("applicationStatus").get
+    val isGis = doc.getAs[BSONDocument]("assistance-details").exists(_.getAs[Boolean]("guaranteedInterview").contains(true))
+    val preferences = doc.getAs[SelectedSchemes]("scheme-preferences").get
+    val applicationBuilder: (List[CubiksTest], Option[PassmarkEvaluation]) => ApplicationReadyForEvaluation =
+      ApplicationReadyForEvaluation(applicationId, applicationStatus, isGis, _, _, preferences)
+    applicationBuilder.curried
+  }
+
+  def passMarkEvaluationReader(passMarkPhase: String, applicationId: String, optDoc: Option[BSONDocument]): PassmarkEvaluation =
+    optDoc.flatMap {_.getAs[BSONDocument]("testGroups")}
+      .flatMap {_.getAs[BSONDocument](passMarkPhase)}
+      .flatMap {_.getAs[PassmarkEvaluation]("evaluation")}
+        .getOrElse(throw PassMarkEvaluationNotFound(applicationId))
+
+
+  def nextApplicationsReadyForEvaluation(currentPassmarkVersion: String, batchSize: Int): Future[List[T]]
+
+  def savePassmarkEvaluation(applicationId: String, evaluation: PassmarkEvaluation,
+                             newApplicationStatus: Option[ApplicationStatus]): Future[Unit] = {
+    val query = BSONDocument("$and" -> BSONArray(
+      BSONDocument("applicationId" -> applicationId),
+      BSONDocument("applicationStatus" -> BSONDocument("$in" -> evaluationApplicationStatuses))
+    ))
+
+    val passMarkEvaluation = BSONDocument("$set" -> BSONDocument(
+      s"testGroups.$phase.evaluation" -> evaluation
+    ).add(
+      if (newApplicationStatus.isDefined) applicationStatusBSON(newApplicationStatus.get) else BSONDocument.empty
```

Comment:
map/getOrElse is a bit nicer IMHO, just because I hate to see a `.get`, but I appreciated it is subjective